base/logging_test.go

```diff
@@ -242,3 +244,140 @@ func TestLogColor(t *testing.T) {
+	assert.Equal(t, "sglogfile.log", path)
+	path = lastComponent("/sglogfile.log")
+	assert.Equal(t, "sglogfile.log", path)
+	path = lastComponent("\\sglogfile.log")
+	assert.Equal(t, "sglogfile.log", path)
+}
+
+func TestLogSyncGatewayVersion(t *testing.T) {
+	out := CaptureOutput(t, LogSyncGatewayVersion)
+	assert.Contains(t, out, LongVersionString)
+	logger := &FileLogger{Enabled: true, logger: &log.Logger{}}
+
+	logger.level = LevelNone
+	assert.True(t, logger.shouldLog(LevelNone))
+	assert.False(t, logger.shouldLog(LevelWarn))
+	assert.False(t, logger.shouldLog(LevelError))
+	assert.False(t, logger.shouldLog(LevelInfo))
+	assert.False(t, logger.shouldLog(LevelDebug))
+	assert.False(t, logger.shouldLog(LevelTrace))
+	out = CaptureOutput(t, LogSyncGatewayVersion)
+	assert.Contains(t, out, LongVersionString)
+
+	logger.level = LevelWarn
+	assert.True(t, logger.shouldLog(LevelNone))
+	assert.True(t, logger.shouldLog(LevelWarn))
+	assert.True(t, logger.shouldLog(LevelError))
+	assert.False(t, logger.shouldLog(LevelInfo))
+	assert.False(t, logger.shouldLog(LevelDebug))
+	assert.False(t, logger.shouldLog(LevelTrace))
+	out = CaptureOutput(t, LogSyncGatewayVersion)
+	assert.Contains(t, out, LongVersionString)
+
+	logger.level = LevelError
+	assert.True(t, logger.shouldLog(LevelNone))
+	assert.False(t, logger.shouldLog(LevelWarn))
+	assert.True(t, logger.shouldLog(LevelError))
+	assert.False(t, logger.shouldLog(LevelInfo))
+	assert.False(t, logger.shouldLog(LevelDebug))
+	assert.False(t, logger.shouldLog(LevelTrace))
+	out = CaptureOutput(t, LogSyncGatewayVersion)
+	assert.Contains(t, out, LongVersionString)
+
+	logger.level = LevelInfo
+	assert.True(t, logger.shouldLog(LevelNone))
+	assert.True(t, logger.shouldLog(LevelWarn))
+	assert.True(t, logger.shouldLog(LevelError))
+	assert.True(t, logger.shouldLog(LevelInfo))
+	assert.False(t, logger.shouldLog(LevelDebug))
+	assert.False(t, logger.shouldLog(LevelTrace))
+	out = CaptureOutput(t, LogSyncGatewayVersion)
+	assert.Contains(t, out, LongVersionString)
+
+	logger.level = LevelDebug
+	assert.True(t, logger.shouldLog(LevelNone))
+	assert.True(t, logger.shouldLog(LevelWarn))
+	assert.True(t, logger.shouldLog(LevelError))
+	assert.True(t, logger.shouldLog(LevelInfo))
+	assert.True(t, logger.shouldLog(LevelDebug))
+	assert.False(t, logger.shouldLog(LevelTrace))
+	out = CaptureOutput(t, LogSyncGatewayVersion)
+	assert.Contains(t, out, LongVersionString)
+
+	logger.level = LevelTrace
+	assert.True(t, logger.shouldLog(LevelNone))
+	assert.True(t, logger.shouldLog(LevelWarn))
+	assert.True(t, logger.shouldLog(LevelError))
+	assert.True(t, logger.shouldLog(LevelInfo))
+	assert.True(t, logger.shouldLog(LevelDebug))
+	assert.True(t, logger.shouldLog(LevelTrace))
+	out = CaptureOutput(t, LogSyncGatewayVersion)
+	assert.Contains(t, out, LongVersionString)
+}
+
+func CaptureOutput(t *testing.T, f func()) string {
+	reader, writer, err := os.Pipe()
+	assert.NoError(t, err)
+	stdout := os.Stdout
```

Comment:
I think we should approach this a little differently than overriding stderr/stdout.

We can make a small change inside the `Consolef` function to:
- Make the output writer a variable we are able to swap out with a bytes.Buffer inside a test
- Correct the default writer to be `os.Stderr` instead of `os.Stdout` (from inside `fmt.Printf()`) (oops! that seems like a me mistake... 😄)

```diff
diff --git a/base/logging.go b/base/logging.go
index 1c1af7a9..146cd4e3 100644
--- a/base/logging.go
+++ b/base/logging.go
@@ -13,6 +13,7 @@ import (
        "context"
        "errors"
        "fmt"
+       "io"
        "math"
        "net/http"
        "net/url"
@@ -517,6 +518,8 @@ func logTo(ctx context.Context, logLevel LogLevel, logKey LogKey, format string,
        }
 }

+var consoleOutput io.Writer = os.Stderr
+
 // Consolef logs the given formatted string and args to the given log level and log key,
 // as well as making sure the message is *always* logged to stdout.
 func Consolef(logLevel LogLevel, logKey LogKey, format string, args ...interface{}) {
@@ -525,7 +528,7 @@ func Consolef(logLevel LogLevel, logKey LogKey, format string, args ...interface
        // If the above logTo didn't already log to stderr, do it directly here
        if !consoleLogger.isStderr || !consoleLogger.shouldLog(logLevel, logKey) {
                format = addPrefixes(format, context.Background(), logLevel, logKey)
-               fmt.Printf(format+"\n", args...)
+               _, _ = fmt.Fprintf(consoleOutput, format+"\n", args...)
        }
 }
```

Reply:
Yeah, this has been committed to the latest revision.